src/Quantum/MP.jl

```diff
@@ -0,0 +1,183 @@
+end
+
+# NOTE does not use optimal contraction path, but "parallel-optimal" which costs x2 more
+# function contractpath(a::TensorNetwork{<:MatrixProductState}, b::TensorNetwork{<:MatrixProductState})
+#     !issetequal(sites(a), sites(b)) && throw(ArgumentError("both tensor networks are expected to have same sites"))
+
+#     b = replace(b, [nameof(outsiteind(b, s)) => nameof(outsiteind(a, s)) for s in sites(a)]...)
+#     path = nameof.(flatten([physicalinds(a), flatten(zip(virtualinds(a), virtualinds(b)))]) |> collect)
+#     inputs = flatten([tensors(a), tensors(b)]) .|> labels
+#     output = Symbol[]
+#     size_dict = merge(size(a), size(b))
+
+#     ContractionPath(path, inputs, output, size_dict)
+# end
+
+# TODO let choose the orthogonality center
+function Base.rand(rng::Random.AbstractRNG, sampler::TNSampler{MatrixProduct{State,Open}})
+    n = sampler.n
+    χ = sampler.χ
+    p = sampler.p
+    T = get(sampler, :eltype, Float64)
+
+    arrays::Vector{AbstractArray{T,N} where {N}} = map(1:n) do i
+        χl, χr = let after_mid = i > n ÷ 2, i = (n + 1 - abs(2i - n - 1)) ÷ 2
+            χl = min(χ, p^(i - 1))
+            χr = min(χ, p^i)
+
+            # swap bond dims after mid and handle midpoint for odd-length MPS
+            (isodd(n) && i == n ÷ 2 + 1) ? (χl, χl) : (after_mid ? (χr, χl) : (χl, χr))
+        end
+
+        # fix for first site
+        i == 1 && ((χl, χr) = (χr, 1))
+
+        # orthogonalize by Gram-Schmidt algorithm
+        A = gramschmidt!(rand(rng, T, χl, χr * p))
+
+        reshape(A, χl, χr, p)
+    end
+
+    # reshape boundary sites
+    arrays[1] = reshape(arrays[1], p, p)
+    arrays[n] = reshape(arrays[n], p, p)
+
+    # normalize state
+    arrays[1] ./= sqrt(p)
+
+    MatrixProduct{State,Open}(arrays; χ = χ)
+end
+
+# TODO let choose the orthogonality center
+# TODO different input/output physical dims
+function Base.rand(rng::Random.AbstractRNG, sampler::TNSampler{MatrixProduct{Operator,Open}})
+    n = sampler.n
+    χ = sampler.χ
+    p = sampler.p
+    T = get(sampler, :eltype, Float64)
+
+    ip = op = p
+
+    arrays::Vector{AbstractArray{T,N} where {N}} = map(1:n) do i
+        χl, χr = let after_mid = i > n ÷ 2, i = (n + 1 - abs(2i - n - 1)) ÷ 2
+            χl = min(χ, ip^(i - 1) * op^(i - 1))
+            χr = min(χ, ip^i * op^i)
+
+            # swap bond dims after mid and handle midpoint for odd-length MPS
+            (isodd(n) && i == n ÷ 2 + 1) ? (χl, χl) : (after_mid ? (χr, χl) : (χl, χr))
+        end
+
+        shape = if i == 1
+            (χr, ip, op)
+        elseif i == n
+            (χl, ip, op)
+        else
+            (χl, χr, ip, op)
+        end
+
+        # orthogonalize by Gram-Schmidt algorithm
+        A = gramschmidt!(rand(rng, T, shape[1], prod(shape[2:end])))
+
+        reshape(A, shape)
+    end
+
+    # normalize
+    ζ = min(χ, ip * op)
+    arrays[1] ./= sqrt(ζ)
+
+    MatrixProductOperator{Open}(arrays; χ = χ)
```

Comment:
```suggestion
    MatrixProduct{Operator,Open}(arrays; χ = χ)
```
This is the fix to the problem spotted by @emapuljak

Reply:
Thanks!

Reply:
Don't resolve the conversation if we haven't commit the change! 😂